// netschoolapi/dist/methods/context/methods/context.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = default_1;
function default_1(client) {
    return __awaiter(this, void 0, void 0, function* () {
        const data = yield client
            .get("context")
            .then((res) => res.json());
        return {
            year: {
                id: data.schoolyear.id,
                gId: data.schoolyear.globalYear.id,
                name: data.schoolyear.name,
                start: new Date(data.schoolyear.startDate),
                end: new Date(data.schoolyear.endDate),
            },
            user: { id: data.user.id, name: data.user.name },
            server: { dateFormat: data.dateFormat, timeFormat: data.timeFormat },
            schoolId: data.schoolId,
        };
    });
}
