// netschoolapi/dist/classes/Client.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.isSecurityWarning = isSecurityWarning;
exports.requestHook = requestHook;
const https_proxy_agent_1 = require("https-proxy-agent");
const web_client_1 = __importDefault(require("@lentryd/web-client"));
function isSecurityWarning(res) {
    return __awaiter(this, void 0, void 0, function* () {
        var _a, _b, _c;
        return (((_b = (_a = res.headers.get("content-type")) === null || _a === void 0 ? void 0 : _a.includes) === null || _b === void 0 ? void 0 : _b.call(_a, "text/html")) &&
            +((_c = res.headers.get("content-length")) !== null && _c !== void 0 ? _c : "") < 1000 &&
            !res.headers.has("filename") &&
            (yield res.clone().text()).includes("/asp/SecurityWarning.asp"));
    });
}
function requestHook(res, url, init) {
    return __awaiter(this, void 0, void 0, function* () {
        if (yield isSecurityWarning(res)) {
            yield this.post("../asp/SecurityWarning.asp", Client.formData({
                at: this.headers.get().at,
                WarnType: 2,
            }));
            return this.request(url, init);
        }
    });
}
__exportStar(require("@lentryd/web-client"), exports);
class Client extends web_client_1.default {
    /**
     * Получает прокси для запросов
     * @returns Прокси
     */
    getProxy() {
        return this.agent.get();
    }
    /**
     * Задает прокси для запросов
     * @param url Ссылка на прокси
     * @returns
     */
    setProxy(url) {
        return this.agent.set(new https_proxy_agent_1.HttpsProxyAgent(url));
    }
}
exports.default = Client;
