// netschoolapi/dist/classes/Diary.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Day_1 = __importDefault(require("./Day"));
class Diary {
    constructor(diary) {
        var _a, _b;
        this.days = (_b = (_a = diary.weekDays) === null || _a === void 0 ? void 0 : _a.map((d) => new Day_1.default(d))) !== null && _b !== void 0 ? _b : [];
        this.termName = diary.termName;
        this.className = diary.className;
        this._end = diary.weekEnd;
        this._start = diary.weekStart;
    }
    get start() {
        return new Date(this._start);
    }
    get end() {
        return new Date(this._end);
    }
    slice({ start, end }) {
        return this.days.filter(({ date }) => date >= start && date < end);
    }
    currentLesson(date) {
        var _a;
        const lessons = [];
        this.days.forEach((d) => lessons.push(...d.lessons));
        return ((_a = lessons.find(({ start: startDate, end: endDate }) => date >= startDate && date < endDate)) !== null && _a !== void 0 ? _a : null);
    }
    toJSON() {
        return {
            days: this.days.map((d) => d.toJSON()),
            endDate: this._end,
            startDate: this._start,
        };
    }
}
exports.default = Diary;
