// netschoolapi/dist/classes/Lesson.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Assignment_1 = __importDefault(require("./Assignment"));
class Lesson {
    constructor(lesson) {
        var _a, _b;
        this.id = lesson.classmeetingId;
        this.subject = lesson.subjectName;
        this._endDate = lesson.day.replace("00:00", lesson.endTime);
        this._startDate = lesson.day.replace("00:00", lesson.startTime);
        this.assignments = (_b = (_a = lesson.assignments) === null || _a === void 0 ? void 0 : _a.map((a) => new Assignment_1.default(a))) !== null && _b !== void 0 ? _b : [];
    }
    get end() {
        return new Date(this._endDate);
    }
    get start() {
        return new Date(this._startDate);
    }
    toJSON() {
        return {
            id: this.id,
            subject: this.subject,
            endDate: this._endDate,
            startDate: this._startDate,
            assignments: this.assignments.map((a) => a.toJSON()),
        };
    }
}
exports.default = Lesson;
