// netschoolapi/dist/methods/grades.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = grades;
const Grades_1 = __importDefault(require("../classes/Grades"));
const dateNum_1 = require("../utils/dateNum");
const checks_1 = require("../utils/checks");
function grades(credentials) {
    return __awaiter(this, void 0, void 0, function* () {
        const { context } = yield checks_1.sessionValid.call(this);
        let { subjectId, start, end, termId, classId, studentId, transport } = credentials;
        // Проверяем существует ли предмет
        if (!context.subjectExists(subjectId))
            throw new Error(`Предмета ${subjectId} не существует`);
        // Проверяем валидность данных
        const termData = checks_1.termIdValid.call(this, termId);
        const classData = checks_1.classIdValid.call(this, classId);
        const studentData = checks_1.studentIdValid.call(this, studentId);
        // Если не указаны даты, то берем текущий учебный год
        const termDates = yield checks_1.termDateValid.call(this, termData.id, start, end);
        start = termDates.start;
        end = termDates.end;
        const [types, htmlText] = yield Promise.all([
            this.assignmentTypes(),
            this.reportFile({
                url: "reports/studentgrades/queue",
                filters: [
                    {
                        filterId: "SID",
                        filterValue: studentData.value,
                    },
                    {
                        filterId: "PCLID_IUP",
                        filterValue: classData.value,
                    },
                    {
                        filterId: "SGID",
                        filterValue: subjectId,
                    },
                    {
                        filterId: "TERMID",
                        filterValue: termData.value,
                    },
                    {
                        filterId: "period",
                        filterValue: (0, dateNum_1.date2JSON)(start) + " - " + (0, dateNum_1.date2JSON)(end),
                    },
                ],
                transport,
            }),
        ]);
        return new Grades_1.default({
            types,
            htmlText,
            hasTerms: context.user.terms.length > 0,
        });
    });
}
