// netschoolapi/dist/NetSchoolApi-safe.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Client_1 = __importStar(require("./classes/Client"));
const logIn_1 = __importDefault(require("./methods/logIn"));
const logOut_1 = __importDefault(require("./methods/logOut"));
const context_1 = __importDefault(require("./methods/context"));
const sessionValid_1 = __importDefault(require("./methods/sessionValid"));
const info_1 = __importDefault(require("./methods/info"));
const photo_1 = __importDefault(require("./methods/photo"));
const diary_1 = __importDefault(require("./methods/diary"));
const assignment_1 = __importDefault(require("./methods/assignment"));
const downloadFile_1 = __importDefault(require("./methods/downloadFile"));
const assignmentTypes_1 = __importDefault(require("./methods/assignmentTypes"));
const scheduleDay_1 = __importDefault(require("./methods/scheduleDay"));
const scheduleWeek_1 = __importDefault(require("./methods/scheduleWeek"));
const reportFile_1 = __importDefault(require("./methods/reportFile"));
const grades_1 = __importDefault(require("./methods/grades"));
const journal_1 = __importDefault(require("./methods/journal"));
const fetch_1 = __importDefault(require("./methods/fetch"));
class NetSchoolApiSafe {
    /**
     * Создание пользователя
     * @/param credentials Данные пользователя
     */
    constructor(credentials) {
        this.context = null;
        this.session = null;
        this.credentials = credentials;
        this.client = new Client_1.default(credentials.origin);
        this.client.onResponse(Client_1.requestHook.bind(this.client));
        this.client.path.set("webapi");
        this.client.headers.set("at", () => { var _a; return ((_a = this.session) === null || _a === void 0 ? void 0 : _a.isValid()) ? this.session.accessToken : undefined; });
    }
    // ⭐️ Пусть будет
    /** Произвольные запросы к сетевому */
    fetch(url, init) {
        return fetch_1.default.call(this, url, init);
    }
    /**
     * Получение прокси
     * @returns Прокси
     */
    getProxy() {
        return this.client.getProxy();
    }
    /**
     * Установка прокси
     * @param url Ссылка на прокси
     */
    setProxy(url) {
        return this.client.setProxy(url);
    }
    // ⭐️ Сессия
    /** Открытие сессии в "Сетевой город. Образование" */
    logIn() {
        return __awaiter(this, void 0, void 0, function* () {
            yield logIn_1.default.call(this);
            if (!this.context)
                this.context = yield context_1.default.call(this);
            return this.session;
        });
    }
    /** Закрытие сессии в "Сетевой город. Образование" */
    logOut() {
        return __awaiter(this, void 0, void 0, function* () {
            yield logOut_1.default.call(this);
            return (this.session = null);
        });
    }
    /** Проверка сессии через API "Сетевой город. Образование"*/
    sessionValid() {
        return sessionValid_1.default.call(this);
    }
    // ⭐️ Пользователь
    /** Информация пользователя */
    info() {
        return info_1.default.call(this);
    }
    /** Фото пользователя */
    photo(credentials) {
        return photo_1.default.call(this, credentials);
    }
    // ⭐️ Дневник
    /** Дневник пользователя*/
    diary(credentials) {
        return diary_1.default.call(this, credentials);
    }
    /** Информация о задание */
    assignment(credentials) {
        return assignment_1.default.call(this, credentials);
    }
    /** Скачивание файла */
    downloadFile(credentials) {
        return downloadFile_1.default.call(this, credentials);
    }
    /** Типы заданий */
    assignmentTypes() {
        return assignmentTypes_1.default.call(this);
    }
    // ⭐️ Расписание
    /** Расписание на день */
    scheduleDay(credentials) {
        return scheduleDay_1.default.call(this, credentials);
    }
    /** Расписание на неделю */
    scheduleWeek(credentials) {
        return scheduleWeek_1.default.call(this, credentials);
    }
    // ⭐️ Отчеты
    reportFile(credentials) {
        return reportFile_1.default.call(this, credentials);
    }
    /** Отчет об успеваемости (по предмету) */
    grades(credentials) {
        return grades_1.default.call(this, credentials);
    }
    /** Отчет об успеваемости (полный) */
    journal(credentials) {
        return journal_1.default.call(this, credentials);
    }
}
exports.default = NetSchoolApiSafe;
